clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -776,17 +776,63 @@ raw2trace_t::do_conversion()
         for (std::thread &thread : threads)
             thread.join();
         for (auto &tdata : thread_data_) {
-            if (!tdata.error.empty())
-                return tdata.error;
-            count_elided_ += tdata.count_elided;
+            if (!tdata->error.empty())
+                return tdata->error;
+            count_elided_ += tdata->count_elided;
         }
     }
+    error = aggregate_and_write_schedule_files();
+    if (!error.empty())
+        return error;
     VPRINT(1, "Reconstructed " UINT64_FORMAT_STRING " elided addresses.\n",
            count_elided_);
     VPRINT(1, "Successfully converted %zu thread files\n", thread_data_.size());
     return "";
 }
 
+std::string
+raw2trace_t::aggregate_and_write_schedule_files()
+{
+    if (serial_schedule_file_ == nullptr && cpu_schedule_file_ == nullptr)
+        return "";
+    std::vector<schedule_entry_t> serial;
+    std::unordered_map<uint64_t, std::vector<schedule_entry_t>> cpu2sched;
+    for (auto &tdata : thread_data_) {
+        serial.insert(serial.end(), tdata->sched.begin(), tdata->sched.end());
+        for (auto &keyval : tdata->cpu2sched) {
+            auto &vec = cpu2sched[keyval.first];
+            vec.insert(vec.end(), keyval.second.begin(), keyval.second.end());
+        }
+    }
+    std::sort(serial.begin(), serial.end(),
+              [](const schedule_entry_t &l, const schedule_entry_t &r) {
+                  return l.timestamp < r.timestamp;
+              });
+    if (serial_schedule_file_ != nullptr) {
+        if (!serial_schedule_file_->write(reinterpret_cast<const char *>(serial.data()),
+                                          serial.size() * sizeof(serial[0])))
+            return "Failed to write to serial schedule file";
+    }
+    if (cpu_schedule_file_ == nullptr)
```

Comment:
Shouldn't this check be at the top of this method perhaps by changing the `&&` to `||` on line 796?

Reply:
Since both are optional the top check only returns immediately of both are empty.  We then have two optional pieces, one for the serial file above this and one for the cpu file which is guarded by this check.